File: src/main/java/net/vanillaoutsider/betterdogs/ai/AvoidHazardsGoal.java
// Copyright (C) 2026 Dasik (Rifaditya) | GNU GPLv3
package net.vanillaoutsider.betterdogs.ai;

import java.util.EnumSet;
import net.minecraft.core.BlockPos;
import net.minecraft.world.entity.ai.goal.Goal;
import net.minecraft.world.entity.animal.Wolf;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.pathfinder.Node;
import net.minecraft.world.level.pathfinder.Path;
import net.vanillaoutsider.betterdogs.registry.BetterDogsGameRules;
import net.vanillaoutsider.betterdogs.util.WolfHazardHelper;

/**
 * AI Goal for wolves to avoid hazardous thermal blocks (Lava, Fire, Magma, Lit Campfires).
 */
public class AvoidHazardsGoal extends Goal {

    private final Wolf wolf;
    private final BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
    private final BlockPos.MutableBlockPos mutableBelowPos = new BlockPos.MutableBlockPos();

    public AvoidHazardsGoal(Wolf wolf) {
        this.wolf = wolf;
        this.setFlags(EnumSet.of(Flag.MOVE));
    }

    @Override
    public boolean canUse() {
        if (wolf.getNavigation().isDone()) {
            return false;
        }

        Level level = wolf.level();
        if (level == null || level.isClientSide()) {
            return false;
        }

        if (!BetterDogsGameRules.getBoolean(level, BetterDogsGameRules.BD_CLIFF_SAFETY, true)) {
            return false;
        }

        Path path = wolf.getNavigation().getPath();
        if (path == null) {
            return false;
        }

        int checkLimit = Math.min(path.getNodeCount(), 8);
        for (int i = 0; i < checkLimit; i++) {
            Node node = path.getNode(i);
            mutablePos.set(node.x, node.y, node.z);
            mutableBelowPos.set(node.x, node.y - 1, node.z);

            if (WolfHazardHelper.isThermalHazard(level.getBlockState(mutablePos)) ||
                WolfHazardHelper.isThermalHazard(level.getBlockState(mutableBelowPos))) {
                return true;
            }
        }

        return false;
    }

    @Override
    public void start() {
        wolf.getNavigation().stop();
    }

    @Override
    public boolean canContinueToUse() {
        return false;
    }
}

File: src/main/kotlin/net/vanillaoutsider/betterdogs/ai/AvoidHazardsGoal.kt
package net.vanillaoutsider.betterdogs.ai

import net.minecraft.core.BlockPos
import net.minecraft.world.entity.ai.goal.Goal
import net.minecraft.world.entity.animal.wolf.Wolf
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.pathfinder.PathType
import java.util.EnumSet

/**
 * AI Goal for wolves to avoid hazardous blocks.
 * Prevents wolves from walking into lava, fire, or off high ledges.
 */
class AvoidHazardsGoal(
    private val wolf: Wolf
) : Goal() {
    
    companion object {
        private const val MAX_SAFE_FALL = 3
        private const val CHECK_RANGE = 2
    }
    
    init {
        flags = EnumSet.of(Flag.MOVE)
    }
    
    override fun canUse(): Boolean {
        // Check if there's a hazard in our movement path
        val path = wolf.navigation.path ?: return false
        
        for (i in 0 until minOf(path.nodeCount, 5)) {
            val node = path.getNode(i)
            val pos = BlockPos(node.x, node.y, node.z)
            
            if (isHazard(pos)) {
                return true
            }
        }
        
        return false
    }
    
    private fun isHazard(pos: BlockPos): Boolean {
        val level = wolf.level()
        val block = level.getBlockState(pos).block
        
        // Check for lava or fire
        if (block == Blocks.LAVA || block == Blocks.FIRE || block == Blocks.SOUL_FIRE) {
            return true
        }
        
        // Check for high fall
        var fallDistance = 0
        var checkPos = pos.below()
        while (fallDistance < 10 && level.getBlockState(checkPos).isAir) {
            fallDistance++
            checkPos = checkPos.below()
        }
        
        if (fallDistance > MAX_SAFE_FALL) {
            return true
        }
        
        return false
    }
    
    override fun start() {
        // Stop current navigation to avoid hazard
        wolf.navigation.stop()
    }
    
    override fun canContinueToUse(): Boolean {
        return false // One-shot goal
    }
}
